op-gateway/src/OPDisputeGameProofService.ts

```diff
@@ -0,0 +1,199 @@
+
+export interface OPProvableBlock {
+  number: number;
+  disputeGameIndex: number;
+}
+
+const DISPUTE_GAME_FACTORY_ABI = [
+  'function gameAtIndex(uint256 _index) external view returns (uint32 gameType_, uint64 timestamp_, address proxy_)',
+  'function gameCount() external view returns (uint256 gameCount_)',
+  'function findLatestGames(uint32 _gameType, uint256 _start, uint256 _n) external view returns (tuple(uint256 index, bytes32 metadata, uint64 timestamp, bytes32 rootClaim, bytes extraData)[] memory games_)',
+];
+
+const FAULT_DISPUTE_GAME_ABI = [
+  // The l2BlockNumber of the disputed output root in the `L2OutputOracle`.
+  'function l2BlockNumber() external view returns (uint256 l2BlockNumber_)',
+  // The output root of the game
+  'function rootClaim() external view returns (bytes32 rootClaim_)',
+  // Status of the game challenging
+  'function status() external view returns (uint8)',
+];
+
+const L2_TO_L1_MESSAGE_PASSER_ADDRESS =
+  '0x4200000000000000000000000000000000000016';
+
+// uint256 index, bytes32 metadata, uint64 timestamp, bytes32 rootClaim, bytes extraData
+interface FindLatestGamesResult {
+  index: bigint;
+  metadata: string;
+  timestamp: bigint;
+  rootClaim: string;
+  extraData: string;
+}
+
+/**
+ * The proofService class can be used to calculate proofs for a given target and slot on the Optimism Bedrock network.
+ * It's also capable of proving long types such as mappings or string by using all included slots in the proof.
+ *
+ */
+export class OPDisputeGameProofService
+  implements IProofService<OPProvableBlock>
+{
+  private readonly disputeGameFactory: Contract;
+  private readonly l1Provider: JsonRpcProvider;
+  private readonly l2Provider: JsonRpcProvider;
+  private readonly helper: EVMProofHelper;
+  private readonly delay: number;
+
+  constructor(
+    l1Provider: JsonRpcProvider,
+    l2Provider: JsonRpcProvider,
+    disputeGameFactoryAddress: string,
+    delay: number
+  ) {
+    this.l1Provider = l1Provider;
+    this.l2Provider = l2Provider;
+    this.helper = new EVMProofHelper(l2Provider);
+    this.delay = delay;
+    this.disputeGameFactory = new Contract(
+      disputeGameFactoryAddress,
+      DISPUTE_GAME_FACTORY_ABI,
+      l1Provider
+    );
+  }
+
+  /**
+   * @dev Returns an object representing a block whose state can be proven on L1.
+   */
+  async getProvableBlock(): Promise<OPProvableBlock> {
+    const respectedGameType = 0;
+
+    /**
+     * Get the latest output from the L2Oracle. We're building the proof with this batch
+     * We go a few batches backwards to avoid errors like delays between nodes
+     *
+     */
+    const gameCount = Number(await this.disputeGameFactory.gameCount());
+
+    const games: FindLatestGamesResult[] =
+      await this.disputeGameFactory.findLatestGames(
+        respectedGameType,
+        gameCount - 1,
+        50
+      );
+
+    const timestampNow = Math.floor(Date.now() / 1000);
```

Comment:
Instead of relying on the local time, we should probably get the latest blocktime.